File: src/test/java/com/example/librarymanagement/service/impl/PublisherServiceImplTest.java
package com.example.librarymanagement.service.impl;

import com.example.librarymanagement.model.entity.Publisher;
import com.example.librarymanagement.repository.PublisherRepository;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class PublisherServiceImplTest {

    @InjectMocks
    private PublisherServiceImpl publisherServiceImpl;

    @Mock
    private PublisherRepository publisherRepository;

    private static Publisher publisher1;
    private static Publisher publisher2;

    @BeforeAll
    public static void init() {
        publisher1 = new Publisher();
        publisher1.setId(1L);
        publisher1.setName("Publisher1");

        publisher2 = new Publisher();
        publisher2.setId(2L);
        publisher2.setName("Publisher2");
    }

    @Test
    public void findAll() {
        when(publisherRepository.findAll()).thenReturn(List.of(publisher1, publisher2));

        assertEquals(2, publisherServiceImpl.getAllPublishers().size());
        verify(publisherRepository, times(1)).findAll();
    }

    @Test
    public void findById() {
        when(publisherRepository.findById(1L)).thenReturn(java.util.Optional.of(publisher1));

        assertEquals(publisher1, publisherServiceImpl.getPublisherById(1L));
        verify(publisherRepository, times(1)).findById(1L);
    }

    @Test
    public void create() {
        when(publisherRepository.save(publisher1)).thenReturn(publisher1);

        assertEquals(publisher1, publisherServiceImpl.createPublisher(publisher1));
        verify(publisherRepository, times(1)).save(publisher1);
    }

    @Test
    public void update() {
        when(publisherRepository.findById(1L)).thenReturn(java.util.Optional.of(publisher1));
        when(publisherRepository.save(publisher1)).thenReturn(publisher1);

        assertEquals(publisher1, publisherServiceImpl.updatePublisher(1L, publisher1));
        verify(publisherRepository, times(1)).findById(1L);
        verify(publisherRepository, times(1)).save(publisher1);
    }
}

File: src/test/java/com/example/librarymanagement/service/impl/PublisherServiceImplTest.kt
package com.example.librarymanagement.service.impl

import com.example.librarymanagement.data.PublisherDataFactory
import com.example.librarymanagement.repository.PublisherRepository
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class PublisherServiceImplTest {
    private val publisherRepository: PublisherRepository = mockk()
    private val publisherService = PublisherServiceImpl(publisherRepository)

    private val id = "1"

    private val publisher = PublisherDataFactory.createPublisher(id)

    @Test
    fun shouldFindAll() {
        //GIVEN
        val expected = listOf(publisher)
        every { publisherRepository.findAll() } returns (listOf(publisher))

        //WHEN
        val actual = publisherService.getAllPublishers()

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { publisherRepository.findAll() }
    }

    @Test
    fun shouldFindById() {
        //GIVEN
        val expected = publisher

        every { publisherRepository.findById(id) } returns publisher

        //WHEN
        val actual = publisherService.getPublisherById(id)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { publisherRepository.findById(id) }
    }

    @Test
    fun shouldCreatePublisher() {
        //GIVEN
        val expected = publisher
        every { publisherRepository.save(publisher) } returns publisher

        //WHEN
        val actual = publisherService.createPublisher(publisher)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { publisherRepository.save(publisher) }
    }

    @Test
    fun shouldUpdatePublisher() {
        //GIVEN
        val excepted = publisher.copy(name = "Updated")
        every { publisherRepository.findById(id) } returns publisher
        every { publisherRepository.save(excepted) } returns excepted

        //WHEN
        val actual = publisherService.updatePublisher(excepted)

        //THEN
        Assertions.assertEquals(excepted, actual)
        verify(exactly = 1) { publisherRepository.findById(id) }
        verify(exactly = 1) { publisherRepository.save(excepted) }
    }
}
